hydra/_internal/core_plugins/basic_sweeper.py

```diff
@@ -152,8 +153,21 @@ def sweep(self, arguments: List[str]) -> Any:
                 f"Validated configs of {len(batch)} jobs in {elapsed:0.2f} seconds, {len(batch)/elapsed:.2f} / second)"
             )
             results = self.launcher.launch(batch, initial_job_idx=initial_job_idx)
+
+            idx = initial_job_idx
+
+            for r in results:
+                try:
+                    _ = r.return_value
+                except Exception as e:
+                    job_failed = True
+                    log.error(f"Job #{idx} failed with exception: {e}")
+                idx += 1
+
             initial_job_idx += len(batch)
             returns.append(results)
+        if job_failed:
+            raise ValueError("Application multirun failed: not all jobs succeeded.")
```

Comment:
1. Use a subclass of HydraException or CompactHydraException.
See the code to learn about the difference.

2. We should probably chain the first exception to the raised one to enable a stack trace. This might require changes to how exceptions are handled in run_and_report().

Reply:
Thanks for the pointers. I've update the exception to be `CompactHydraException` and chained the first exception. 

Here is what the stack trace looks like  as of (10f857d4380c0a89d19f3b6a915430a283863814)

### Without HYDRA_FULL_ERROR=1
```
$  python tests/test_apps/app_can_fail/my_app.py +divisor=1,0,0 -m
[2021-03-17 11:27:35,688][HYDRA] Launching 3 jobs locally
[2021-03-17 11:27:35,689][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-17 11:27:35,762][HYDRA]        #1 : +divisor=0
[2021-03-17 11:27:35,846][HYDRA]        #2 : +divisor=0
[2021-03-17 11:27:35,921][HYDRA] Job #1 failed.
[2021-03-17 11:27:35,921][HYDRA] Job #2 failed.
hydra.errors.FailedMultirunException: Multirun failed, failed job: #1,#2

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```


### With HYDRA_FULL_ERROR=1
```
$ HYDRA_FULL_ERROR=1  python tests/test_apps/app_can_fail/my_app.py +divisor=1,0,0 -m
[2021-03-17 11:25:40,050][HYDRA] Launching 3 jobs locally
[2021-03-17 11:25:40,050][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-17 11:25:40,133][HYDRA]        #1 : +divisor=0
[2021-03-17 11:25:40,208][HYDRA]        #2 : +divisor=0
[2021-03-17 11:25:40,283][HYDRA] Job #1 failed.
[2021-03-17 11:25:40,283][HYDRA] Job #2 failed.
hydra.errors.FailedMultirunException: Multirun failed, failed job: #1,#2

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 14, in <module>
    my_app()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/main.py", line 29, in decorated_main
    _run_hydra(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 369, in _run_hydra
    run_and_report(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 213, in run_and_report
    raise ex
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 210, in run_and_report
    return func()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 370, in <lambda>
    lambda: hydra.multirun(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/hydra.py", line 126, in multirun
    return sweeper.sweep(arguments=task_overrides)
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/core_plugins/basic_sweeper.py", line 174, in sweep
    raise first_exception from FailedMultirunException(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/core_plugins/basic_sweeper.py", line 162, in sweep
    _ = r.return_value
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 197, in return_value
    raise self._return_value
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 131, in run_job
    ret._return_value = task_function(task_cfg)
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero
```

Reply:
This looks good. Can you verify that it looks similar when using a more sophisticated launcher? (not the basic one).